In [121]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_curve, auc
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import train_test_split

In [122]:
df_features = pd.read_csv("training_set_features.csv");
df_labels = pd.read_csv("training_set_labels.csv")


# Exploratory Data Analysis(EDA)


    1.Here I processed and cleant the raw data .
    2.there are 35 features .
    3.Firstly I seperated the Categorical and Numerical Variables columns
    4.Then I handeled the missing Values of Categorical variable by assigning the nan values as  missing category
    5.Then I handeled the missing Values of Numerical variables by using the median
    6.After that I scaled the data accordingly


In [123]:
df_features

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0                  0          1.0            0.0                        0.0   
1                  1          3.0            2.0                        0.0   
2                  2          1.0            1.0                        0.0   
3                  3          1.0            1.0                        0.0   
4                  4          2.0            1.0                        0.0   
...              ...          ...            ...                        ...   
26702          26702          2.0            0.0                        0.0   
26703          26703          1.0            2.0                        0.0   
26704          26704          2.0            2.0                        0.0   
26705          26705          1.0            1.0                        0.0   
26706          26706          0.0            0.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                       0.0                   0.0                    0.0   
1                       1.0                   0.0                    1.0   
2                       1.0                   0.0                    0.0   
3                       1.0                   0.0                    1.0   
4                       1.0                   0.0                    1.0   
...                     ...                   ...                    ...   
26702                   1.0                   0.0                    0.0   
26703                   1.0                   0.0                    1.0   
26704                   1.0                   1.0                    1.0   
26705                   0.0                   0.0                    0.0   
26706                   1.0                   0.0                    0.0   

       behavioral_large_gatherings  behavioral_outside_home  \
0                              0.0                      1.0   
1                              0.0                      1.0   
2                              0.0                      0.0   
3                              1.0                      0.0   
4                              1.0                      0.0   
...                            ...                      ...   
26702                          0.0                      1.0   
26703                          0.0                      0.0   
26704                          1.0                      0.0   
26705                          0.0                      0.0   
26706                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
0                        1.0  ...              Below Poverty     Not Married   
1                        1.0  ...              Below Poverty     Not Married   
2                        0.0  ...  <= $75,000, Above Poverty     Not Married   
3                        0.0  ...              Below Poverty     Not Married   
4                        1.0  ...  <= $75,000, Above Poverty         Married   
...                      ...  ...                        ...             ...   
26702                    0.0  ...  <= $75,000, Above Poverty     Not Married   
26703                    0.0  ...  <= $75,000, Above Poverty     Not Married   
26704                    1.0  ...                        NaN     Not Married   
26705                    NaN  ...  <= $75,000, Above Poverty         Married   
26706                    0.0  ...  <= $75,000, Above Poverty         Married   

       rent_or_own   employment_status  hhs_geo_region  \
0              Own  Not in Labor Force        oxchjgsf   
1             Rent            Employed        bhuqouqj   
2              Own            Employed        qufhixun   
3             Rent  Not in Labor Force        lrircsnp   
4              Own            Employed        qufhixun   
...            ...                 ...             ...   
26702          Own  N

In [124]:
df_labels

respondent_id  xyz_vaccine  seasonal_vaccine
0                  0            0                 0
1                  1            0                 1
2                  2            0                 0
3                  3            0                 1
4                  4            0                 0
...              ...          ...               ...
26702          26702            0                 0
26703          26703            0                 0
26704          26704            0                 1
26705          26705            0                 0
26706          26706            0                 0

[26707 rows x 3 columns]

## Extracting the Categorical and Numerical Features


In [125]:
categorical_feac = [ feature for feature in df_features.columns if df_features[feature].dtypes == "O"]
numerical_feac = [ feature for feature in df_features.columns if df_features[feature].dtypes != "O"]

In [126]:
categorical_feac

['age_group',
 'education',
 'race',
 'sex',
 'income_poverty',
 'marital_status',
 'rent_or_own',
 'employment_status',
 'hhs_geo_region',
 'census_msa',
 'employment_industry',
 'employment_occupation']

## Checking Null Values for Categorical Features

In [127]:
df_features[categorical_feac].isnull().sum()

age_group                    0
education                 1407
race                         0
sex                          0
income_poverty            4423
marital_status            1408
rent_or_own               2042
employment_status         1463
hhs_geo_region               0
census_msa                   0
employment_industry      13330
employment_occupation    13470
dtype: int64

## Checking Null Values for Numerical Features

In [128]:
df_features[numerical_feac].isnull().sum()

respondent_id                      0
xyz_concern                       92
xyz_knowledge                    116
behavioral_antiviral_meds         71
behavioral_avoidance             208
behavioral_face_mask              19
behavioral_wash_hands             42
behavioral_large_gatherings       87
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_xyz                 2160
doctor_recc_seasonal            2160
chronic_med_condition            971
child_under_6_months             820
health_worker                    804
health_insurance               12274
opinion_xyz_vacc_effective       391
opinion_xyz_risk                 388
opinion_xyz_sick_from_vacc       395
opinion_seas_vacc_effective      462
opinion_seas_risk                514
opinion_seas_sick_from_vacc      537
household_adults                 249
household_children               249
dtype: int64

# Labelling the Categorical features using the Standard Label

In [129]:
label_en = LabelEncoder()


for i in categorical_feac:
    label = label_en.fit_transform(df_features[i])
    df_features.drop(i, axis=1, inplace=True)
    df_features[i] = label

In [130]:
df_features[categorical_feac].isnull().sum()

age_group                0
education                0
race                     0
sex                      0
income_poverty           0
marital_status           0
rent_or_own              0
employment_status        0
hhs_geo_region           0
census_msa               0
employment_industry      0
employment_occupation    0
dtype: int64

In [131]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import RobustScaler
from sklearn.preprocessing import OneHotEncoder

# Creating Pipelines to scale and replace null values both for Categorical and Numerical Variables

In [132]:
numerical_transformer = Pipeline(steps = [
    ('scaler', RobustScaler()),   
    ('imputer', SimpleImputer(strategy='mean')),
])

In [133]:
df_features[numerical_feac] = pd.DataFrame(numerical_transformer.fit_transform(df_features[numerical_feac]),columns=numerical_feac)

In [134]:
df_features[numerical_feac].isnull().sum()

respondent_id                  0
xyz_concern                    0
xyz_knowledge                  0
behavioral_antiviral_meds      0
behavioral_avoidance           0
behavioral_face_mask           0
behavioral_wash_hands          0
behavioral_large_gatherings    0
behavioral_outside_home        0
behavioral_touch_face          0
doctor_recc_xyz                0
doctor_recc_seasonal           0
chronic_med_condition          0
child_under_6_months           0
health_worker                  0
health_insurance               0
opinion_xyz_vacc_effective     0
opinion_xyz_risk               0
opinion_xyz_sick_from_vacc     0
opinion_seas_vacc_effective    0
opinion_seas_risk              0
opinion_seas_sick_from_vacc    0
household_adults               0
household_children             0
dtype: int64

In [135]:
categorical_transformer = Pipeline(steps=[
    ('scaler', RobustScaler()),
])

df_features[categorical_feac] = pd.DataFrame(categorical_transformer.fit_transform(df_features[categorical_feac]),columns=categorical_feac)

In [136]:
df_features.isnull().sum()

respondent_id                  0
xyz_concern                    0
xyz_knowledge                  0
behavioral_antiviral_meds      0
behavioral_avoidance           0
behavioral_face_mask           0
behavioral_wash_hands          0
behavioral_large_gatherings    0
behavioral_outside_home        0
behavioral_touch_face          0
doctor_recc_xyz                0
doctor_recc_seasonal           0
chronic_med_condition          0
child_under_6_months           0
health_worker                  0
health_insurance               0
opinion_xyz_vacc_effective     0
opinion_xyz_risk               0
opinion_xyz_sick_from_vacc     0
opinion_seas_vacc_effective    0
opinion_seas_risk              0
opinion_seas_sick_from_vacc    0
household_adults               0
household_children             0
age_group                      0
education                      0
race                           0
sex                            0
income_poverty                 0
marital_status                 0
rent_or_ow

In [137]:
X_train = df_features.drop(["respondent_id"],axis=1)
Y_train = df_labels.drop(["respondent_id"],axis=1)

In [138]:
X_train, X_test, y_train, y_test = train_test_split(
    X_train, Y_train, test_size=0.2, random_state=42)

# Using ML Algorithms


here I used 2 major Classification methods 
        1.Logistic Regression
        2.SVM

In [139]:
X_train_cpy = X_train
X_test_cpy = X_test
X_train

xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
24706          0.0            0.0                        0.0   
5393           0.0            0.0                        1.0   
20898         -2.0            0.0                        0.0   
3429          -1.0            0.0                        0.0   
8731          -1.0           -1.0                        0.0   
...            ...            ...                        ...   
21575          0.0            0.0                        0.0   
5390          -1.0            0.0                        0.0   
860            0.0            0.0                        0.0   
15795          0.0            0.0                        0.0   
23654          1.0            0.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
24706                   0.0                   0.0                    0.0   
5393                    0.0                   1.0                    0.0   
20898                  -1.0                   0.0                    0.0   
3429                   -1.0                   0.0                   -1.0   
8731                   -1.0                   0.0                    0.0   
...                     ...                   ...                    ...   
21575                  -1.0                   0.0                    0.0   
5390                   -1.0                   0.0                    0.0   
860                     0.0                   0.0                   -1.0   
15795                   0.0                   0.0                    0.0   
23654                  -1.0                   0.0                    0.0   

       behavioral_large_gatherings  behavioral_outside_home  \
24706                          0.0                      1.0   
5393                           1.0                      1.0   
20898                          0.0                      1.0   
3429                           0.0                      0.0   
8731                           0.0                      0.0   
...                            ...                      ...   
21575                          0.0                      0.0   
5390                           0.0                      0.0   
860                            0.0                      0.0   
15795                          0.0                      0.0   
23654                          0.0                      1.0   

       behavioral_touch_face  doctor_recc_xyz  ...  race  sex  income_poverty  \
24706                    0.0         0.000000  ...   0.0  0.0             0.5   
5393                     0.0         0.000000  ...   0.0  1.0             0.0   
20898                   -1.0         0.000000  ...   0.0  1.0             1.0   
3429                    -1.0         0.000000  ...  -1.0  0.0             1.0   
8731                    -1.0         0.000000  ...  -3.0  0.0            -0.5   
...                      ...              ...  ...   ...  ...             ...   
21575                    0.0         0.000000  ...   0.0  1.0             0.0   
5390                     0.0         0.220312  ...   0.0  1.0            -0.5   
860                      0.0         0.000000  ...   0.0  0.0            -0.5   
15795                    0.0         1.000000  ...  -3.0  0.0             0.0   
23654                    0.0         0.000000  ...  -3.0  1.0            -0.5   

       marital_status  rent_or_own  employment_status  hhs_geo_region  \
24706             0.0          1.0                2.0            0.25   
5393              1.0          0.0                0.0            0.25   
20898             0.0          2.0                0.0            1.00   
3429              1.0          0.0                0.0           -1.00   
8731              1.0          0.0                1.0            0.50   
...               ...          ...                ...             ...   
21575             1.0          0.0                1.0            1.00   
5390       

In [140]:


feature_sel_model = SelectFromModel(LogisticRegression()) # remember to set the seed, the random state in this function
feature_sel_model.fit(X_train, y_train["xyz_vaccine"])
selected_feat = X_train.columns[(feature_sel_model.get_support())]

selected_feat

Index(['doctor_recc_xyz', 'doctor_recc_seasonal', 'health_worker',
       'health_insurance', 'opinion_xyz_vacc_effective', 'opinion_xyz_risk',
       'opinion_seas_risk', 'age_group'],
      dtype='object')

In [141]:
X_train=X_train
X_test = X_test

In [142]:
model1 = LogisticRegression()

model1.fit(X_train, y_train["xyz_vaccine"])

LogisticRegression()

In [143]:

feature_sel_model = SelectFromModel(LogisticRegression()) # remember to set the seed, the random state in this function
feature_sel_model.fit(X_train, y_train["seasonal_vaccine"])
selected_feat2 = X_train.columns[(feature_sel_model.get_support())]

selected_feat2

Index(['doctor_recc_xyz', 'doctor_recc_seasonal', 'health_worker',
       'health_insurance', 'opinion_seas_vacc_effective', 'opinion_seas_risk',
       'opinion_seas_sick_from_vacc', 'age_group'],
      dtype='object')

In [144]:
X_train2 = X_train
X_test2= X_test

In [145]:
model2 = LogisticRegression()
model2.fit(X_train2, y_train["seasonal_vaccine"])

LogisticRegression()

In [146]:
df_test = pd.read_csv("test_set_features.csv")

In [147]:
df_test_id = df_test["respondent_id"]

In [148]:
df_test.isnull().sum()

respondent_id                      0
xyz_concern                       85
xyz_knowledge                    122
behavioral_antiviral_meds         79
behavioral_avoidance             213
behavioral_face_mask              19
behavioral_wash_hands             40
behavioral_large_gatherings       72
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_xyz                 2160
doctor_recc_seasonal            2160
chronic_med_condition            932
child_under_6_months             813
health_worker                    789
health_insurance               12228
opinion_xyz_vacc_effective       398
opinion_xyz_risk                 380
opinion_xyz_sick_from_vacc       375
opinion_seas_vacc_effective      452
opinion_seas_risk                499
opinion_seas_sick_from_vacc      521
age_group                          0
education                       1407
race                               0
sex                                0
income_poverty                  4497
m

In [149]:
label_en = LabelEncoder()


for i in categorical_feac:
    label = label_en.fit_transform(df_test[i])
    df_test.drop(i, axis=1, inplace=True)
    df_test[i] = label

In [150]:
df_test[categorical_feac].isnull().sum()

age_group                0
education                0
race                     0
sex                      0
income_poverty           0
marital_status           0
rent_or_own              0
employment_status        0
hhs_geo_region           0
census_msa               0
employment_industry      0
employment_occupation    0
dtype: int64

In [151]:
df_test[numerical_feac] = pd.DataFrame(numerical_transformer.fit_transform(df_test[numerical_feac]),columns=numerical_feac)

Checking the Null Values for Numerical Variables again

In [152]:
df_test[numerical_feac].isnull().sum()

respondent_id                  0
xyz_concern                    0
xyz_knowledge                  0
behavioral_antiviral_meds      0
behavioral_avoidance           0
behavioral_face_mask           0
behavioral_wash_hands          0
behavioral_large_gatherings    0
behavioral_outside_home        0
behavioral_touch_face          0
doctor_recc_xyz                0
doctor_recc_seasonal           0
chronic_med_condition          0
child_under_6_months           0
health_worker                  0
health_insurance               0
opinion_xyz_vacc_effective     0
opinion_xyz_risk               0
opinion_xyz_sick_from_vacc     0
opinion_seas_vacc_effective    0
opinion_seas_risk              0
opinion_seas_sick_from_vacc    0
household_adults               0
household_children             0
dtype: int64

In [153]:
df_test1 = df_test.drop(["respondent_id"],axis=1)


In [154]:
from sklearn.metrics import roc_auc_score
val_preds_xyz = model1.predict_proba(X_test)[:, 1]
val_preds_seasonal = model2.predict_proba(X_test2)[:, 1]


roc_auc_xyz = roc_auc_score(y_test['xyz_vaccine'], val_preds_xyz)

roc_auc_seasonal = roc_auc_score(y_test['seasonal_vaccine'], val_preds_seasonal)
print(f'ROC AUC for xyz_vaccine: {roc_auc_xyz}')
print(f'ROC AUC for seasonal_vaccine: {roc_auc_seasonal}')



print(f'Mean ROC AUC: {(roc_auc_xyz + roc_auc_seasonal) / 2}')

ROC AUC for xyz_vaccine: 0.8324219045457982
ROC AUC for seasonal_vaccine: 0.8504296102608003
Mean ROC AUC: 0.8414257574032993


# ROC AUC Score Logistic Regression:


        1. for model1 ROC AUC Score is  0.8324219045457982
        2. for model2 ROC AUC Score is  0.8504296102608003
        3. Avg Score is  0.8414257574032993

# SVM (try)

    This is just for experimentation 

In [155]:
from sklearn.svm import SVC
model4 = SVC(C=100,gamma=0.01,probability= True,kernel="rbf")
model3 = SVC(C=100,gamma=0.01,probability= True,kernel="rbf")
model4.fit(X_train_cpy,y_train["xyz_vaccine"])
model3.fit(X_train_cpy,y_train["seasonal_vaccine"])


SVC(C=100, gamma=0.01, probability=True)

In [156]:
val_preds_SVM = model4.predict_proba(X_test)[:, 1]
val_preds_SVM2 = model3.predict_proba(X_test)[:, 1]

roc_auc_SVM = roc_auc_score(y_test['xyz_vaccine'], val_preds_SVM)
roc_auc_SVM2 = roc_auc_score(y_test['seasonal_vaccine'], val_preds_SVM2)

print(f'Mean ROC AUC 1: {(roc_auc_SVM)}')
print(f'Mean ROC AUC 2: {(roc_auc_SVM2)}')

Mean ROC AUC 1: 0.8287125280488113
Mean ROC AUC 2: 0.8543996400709527


# Random Forest Classifier:

 Here I tried to implement the Random Forest from the  scklitlearn library

In [157]:
from sklearn.ensemble import RandomForestClassifier

In [158]:
model5 = RandomForestClassifier(n_estimators = 100)
model6 = RandomForestClassifier(n_estimators = 100) 

In [159]:
model5.fit(X_train2,y_train["xyz_vaccine"])
model6.fit(X_train2,y_train["seasonal_vaccine"])

RandomForestClassifier()

In [160]:
from sklearn.metrics import roc_auc_score
val_preds_rdmf = model5.predict_proba(X_test)[:, 1]
val_preds_rdmf2 = model6.predict_proba(X_test)[:, 1]


roc_auc_rdmf = roc_auc_score(y_test["xyz_vaccine"], val_preds_rdmf)
roc_auc_rdmf2 = roc_auc_score(y_test["seasonal_vaccine"], val_preds_rdmf2)
print(f'ROC AUC for xyz_vaccine: {roc_auc_rdmf}')
print(f'ROC AUC for seasonal_vaccine: {roc_auc_rdmf2}')


ROC AUC for xyz_vaccine: 0.862072754624377
ROC AUC for seasonal_vaccine: 0.8583089149192029


#  ROC AUC Score Summary (All Models):
       for Logistis Regression:
           model 1 (xyz_vaccine) : 0.8324219045457982
           model 2(Seasonal vaccine) : 0.8504296102608003
           avg : 0.8414257574032993
           
           
       for SVM :
           model 3 (xyz_vaccine): 0.8287291262217517
           model 4(Seasonal vaccine): 0.854396535287766
       
       
       
       for Random Forest Classifier :
           model 5(xyz_vaccine) : 0.8638627310087487
           model 6(seasonal_vaccine): 0.856239647488562
           
           
           
           
           
       So the best probable Classification models for this data set are model 5 and model6

In [161]:
test_xyz = model5.predict_proba(df_test1)[:, 1]
test_seasonal = model6.predict_proba(df_test1)[:, 1]

In [162]:
submission = pd.DataFrame({
    'respondent_id': df_test_id,
    'xyz_vaccine': test_xyz,
    'seasonal_vaccine': test_seasonal
})


# Saving submission to CSV
submission.to_csv('final_submission.csv', index=False)